src/excmds.ts

```diff
@@ -1706,11 +1706,66 @@ export async function tabprev(increment = 1) {
     // Proper way:
     // return tabIndexSetActive((await activeTab()).index - increment + 1)
     // Kludge until https://bugzilla.mozilla.org/show_bug.cgi?id=1504775 is fixed:
-    return browser.tabs.query({ currentWindow: true, hidden: false }).then(tabs => {
-        tabs = tabs.sort((t1, t2) => t1.index - t2.index)
-        let prevTab = (tabs.findIndex(t => t.active) - increment + tabs.length) % tabs.length
-        return browser.tabs.update(tabs[prevTab].id, { active: true })
-    })
+
+    var hasTST = false
+    try {
+        // Not sure why this is an error
+        await browser.management.get("treestyletab@piro.sakura.ne.jp")
+        hasTST = true
+    } catch (e) {
+        hasTST = false
+    }
+
+    if (config.get("treestyletabintegration") && hasTST) {
+        // Ok so this entire piece here is really inefficient, it looks up all tabs, gets the active tab id, gets all tabs again (this time as a tree), iterates through all those tabs flattening them into an array, and then iterates over them once more to find the index of the active tab in that flattened array. This has a lot of room for improvement in the future.
+        // Find the current TAB id
+        let activeTabId = (await activeTab()).id
+        // Get the whole tab tree
+        let tabTree = await browser.runtime.sendMessage("treestyletab@piro.sakura.ne.jp", {
+            type: "get-tree",
+            tabs: "*",
+        })
+        // Convert the Tree to a flat array
```

Comment:
You might check config.ts for similar functions - I recall writing a tree walker for that.